docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,324 @@
+```http request
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: br, gzip 
+```
+
+### Compression Algorithms
+Given that Kotlin is multiplatform there should be multiple implementations of each compression algorithm. Currently gzip is the only supported compression algorithm by the server side and the only one supported for
+the SDK. Only gzip for JVM is supported specifically. This is done by using the following classes from the Java standard library: `GZIPInputStream` and `GZIPOutputStream`. The other KMP targets are not implemented, and are planned to be implemented using the CRT.
+
+New algorithms can be added as they become supported using the compression algorithm interface and by adding them to the list of supported algorithms. The same goes for non JVM KMP targets, these can be added by filling in the `actual` declarations for the target.
+
+#### Interfaces
+This interface is used to represent a compression algorithm.
+
+```kotlin
+/**
+ * Represents a compression algorithm to be used for compressing request payloads on qualifying operations
+ */
+public interface CompressionAlgorithm {
+    /**
+     * The ID of the compression algorithm
+     */
+    public val id: String
+    
+    /**
+     * Compresses a payload
+     */
+    public suspend fun compress(stream: HttpBody): HttpBody
+}
+```
+
+The `id` is used to match a supported `CompressionAlgorithm` to a requested compression algorithm from the trait. The `compress` function is used to compress HTTP bodies.
+
+#### Compression
+Gzip JVM compression implementation:
+
+```kotlin
+/**
+
+Plan on writing code to convert SourceContent & ChannelContent to  
+`java.io.OutputStream` to use `GZIPOutputStream` in similar fashion as above.
+`GZIPOutputStream` takes in a `java.io.OutputStream` which `ByteArrayOutputStream` is.
+
+Could also use function to convert stream to byte array but chances are will run out
+of memory doing so. Body will probably be large if using streaming trait on it.
+
+If that code to convert the types is not possible for some reason then will have to 
+handwrite gzip implementation.
+ 
+**/
+```
+
+### Codegen
+Middleware registration has the following steps:
+
+* A Kotlin integration will look at a service model
+* Determine if the service has an operation with the `requestCompression` trait
+* If so middleware is registered  for the service
+* The middleware will look at each operation in the service individually
+* The middleware will then determine if an operation has the `requestCompression` trait
+* If so then it'll add an interceptor to the operation
+
+```kotlin
+/**
+ * Middleware that enables compression of payloads for HTTP requests
+ */
+class RequestCompressionTrait : KotlinIntegration {
+
+    // Will determine if middleware is registered for service
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean = model
+        .shapes<OperationShape>()
```

Comment:
**nit**: We can address this in review but as an FYI we should not directly iterate operation shapes like this, we always need to consult the service closure.